docs: add explainer on performance budgets (#10542)
GoogleChrome/lighthouse: docs/performance-budgets.md

# Performance Budgets (budget.json)

> Use a performance budget to prevent site performance from regressing over time.

Reviewer: in the context of lighthouse we might want a more specific overview explanation, maybe something like

```suggestion
Use a performance budget to assert a custom file size or count limit per resource type and thresholds for each performance metric.
```

? feel free to reword just something unique to budgets would be nice :)

Reviewer 2: related: patrick should we mention interop w/ LHCI here?

Reviewer: yeah a good chance to plug the budgets interop there too :)

## Usage

Performance budgets are supported in the CLI version of lighthouse. To use:
1. Create a `budget.json` file.

> 2. When running Lighthouse from the command line, pass the `--budget-path` (or `--budgetPath`) flag followed by the path to budget file in order to calculate whenever a category is over budget.

Reviewer 2: ```suggestion
2. When running Lighthouse from the command line, pass the `--budget-path` flag followed by the path to budget file in order to calculate whenever a category is over budget.
```

don't think that's necessary

```json
[
  {
    "path": "/*",
    "options": {
       "firstPartyHostnames": ["*.my-site.com", "my-site.cdn.com"]
    },
    "timings": [
      {
        "metric": "interactive",
        "budget": 5000
      },
      {
        "metric": "first-meaningful-paint",
        "budget": 2000
      }
    ],
    "resourceSizes": [
      {
        "resourceType": "total",
        "budget": 500
      },
      {
        "resourceType": "script",
        "budget": 150
      }
    ],
    "resourceCounts": [
      {
        "resourceType": "third-party",
        "budget": 100
      }
    ]
  },
  {
    "options": {
       "firstPartyHostnames": ["*.my-site.com", "my-site.cdn.com"]
    },
    "path": "/checkout",
    "resourceSizes": [
      {
        "resourceType": "script",
        "budget": 200
      }
    ]
  }
]
``` 

## Further Explanation

### Timing Budgets

_Lighthouse 6 & up_

Use the optional `timings` property to define budgets for time-based performance metrics. In this context, budgets are defined in  milliseconds.

```json
"timings": [
   {
         "metric": "interactive",
         "budget": 5000
   }
]
```

Supported timing metrics:

> - `first-contentful-paint`

Reviewer: we should add `largest-contentful-paint` support to this list before 6.0 final :)

Reviewer 2: Let's just add it now? + CLS. If you don't like that we could at least make an issue now.

```json
"resourceCounts": [
   {
      "resourceType": "script",
      "budget": 10
   }
]
```

Budgets can be set for the follow resource types.

*   `document`
*   `font`
*   `image`
*   `media`
*   `other`
*   `script`
*   `stylesheet`
*   `third-party`
*   `total`

### Using the `path` property

_Lighthouse 5.3 & up_

The `path` property indciates the pages that a budget applies to. This string should follow the [robots.txt](https://developers.google.com/search/reference/robots_txt#examples-of-valid-robotstxt-urls) format.

If `path` is not supplied, a budget will apply to all pages.

If a page's URL path matches the `path` property of more than one budget in `budget.json`, then the last matching budget will be applied. As a result, global budgets (e.g. `"path": "/*"`) should be listed first in `budget.json`, followed by the budgets that override the global budget (e.g. `"path": "/blog"`). 

Examples:

Match all URL paths.

`"path": "/"` (This is equivalent to writing `"path": "/*"`)

Match all URL paths starting with `/articles`.

`"path": "/articles"`

> Match URL paths within the `uk/` directory and ending with `shopping-cart`.

Reviewer: ```suggestion
Match URL paths within the `uk/` directory and ending with `/shopping-cart`.
```

right? or maybe I'm misremembering the logic for `/`

aside: would `/uk/shopping-cart` match with this pattern? I *think* the answer is no, so might want to tweak the example so the explanation matches but please ignore if I'm wrong

# Performance Budgets (budget.json)

> Use a performance budget to assert thresholds for performance metrics. `budget.json` supports three types of budgets:

Reviewer: any opinion here @khempenius or should I feel free to merge? :)

> Use a performance budget to assert thresholds for performance metrics. `budget.json` supports three types of budgets:

Author: ```suggestion
Use a [performance budget](https://addyosmani.com/blog/performance-budgets/) to assert thresholds for performance metrics. `budget.json` supports three types of budgets:
```

> Use a performance budget to assert thresholds for performance metrics. `budget.json` supports three types of budgets:

Author: Bah, sorry for the delay - in my notifications it looked like this had been merged before I responded 🤦‍♀️.

I have a deck that I'd like to link here, but it needs to clean up before release - so this link should be good for now.

> Use a [performance budget](https://addyosmani.com/blog/performance-budgets/) to assert thresholds for performance metrics. `budget.json` supports three types of budgets:

Author: ```suggestion
Use a [performance budget](https://web.dev/performance-budgets-101/) to assert thresholds for performance metrics. `budget.json` supports three types of budgets:
```
